Bump Go version, deps, fix some linter issues...

kolyshkin opened the pull request:
This started as a bump of some CI deps but quickly got out of hand 😬 

See individual commits for details. High level overview:
- ci: bump ubuntu to 24.04 (see https://github.com/actions/runner-images/issues/11101);
- ~~go.mod: bump go to 1.21, drop min/max functions;~~
- ci: add go 1.19, 1.23, 1.24, drop go 1.21, 1.22 from test matrix;
- fix or silence new gosec warnings (these commits need to be reviewed carefully);
- ci: bump golangci-lint to v1.64, fix golangci-lint jobs caching;
- ci: add 32-bit tests;
- ci: add `all-done` job;
- go.mod: bump to `golang.org/x/sys v0.1.0`;
- ci: show avc denials after tests.

kolyshkin commented:
Oops, apparently most tests are skipped (because, of course, Ubuntu does not have selinux).

```
=== RUN   TestSetFileLabel
    selinux_linux_test.go:16: SELinux not enabled, skipping.
--- SKIP: TestSetFileLabel (0.00s)
=== RUN   TestKVMLabels
    selinux_linux_test.go:83: SELinux not enabled, skipping.
--- SKIP: TestKVMLabels (0.00s)
=== RUN   TestInitLabels
    selinux_linux_test.go:104: SELinux not enabled, skipping.
--- SKIP: TestInitLabels (0.00s)
=== RUN   TestSELinux
    selinux_linux_test.go:136: SELinux not enabled, skipping.
--- SKIP: TestSELinux (0.00s)
=== RUN   TestSetEnforceMode
    selinux_linux_test.go:183: SELinux not enabled, skipping.
--- SKIP: TestSetEnforceMode (0.00s)
=== RUN   TestCanonicalizeContext
    selinux_linux_test.go:206: SELinux not enabled, skipping.
--- SKIP: TestCanonicalizeContext (0.00s)
=== RUN   TestFindSELinuxfsInMountinfo
    selinux_linux_test.go:270: found "/sys/fs/selinux"
    selinux_linux_test.go:270: found "/root/chroot/selinux"
    selinux_linux_test.go:270: found ""
--- PASS: TestFindSELinuxfsInMountinfo (0.00s)
=== RUN   TestSecurityCheckContext
    selinux_linux_test.go:279: SELinux not enabled, skipping.
--- SKIP: TestSecurityCheckContext (0.00s)
=== RUN   TestClassIndex
    selinux_linux_test.go:304: SELinux not enabled, skipping.
--- SKIP: TestClassIndex (0.00s)
=== RUN   TestComputeCreateContext
    selinux_linux_test.go:324: SELinux not enabled, skipping.
....
```

Need to switch to a hosted Fedora or something like that. But that's a separate issue.

Let this be draft for now.

kolyshkin commented:
> Oops, apparently most tests are skipped (because, of course, Ubuntu does not have selinux).

Of course [we knew it for a long time](https://github.com/opencontainers/selinux/pull/194#issuecomment-1300420091), we just forgot :)

And since https://github.com/actions/runner-images/issues/2307 is quite old, I filed a new one: https://github.com/actions/runner-images/issues/10802

rhatdan commented:
LGTM, Are the SELinux tests running on Ubuntu now?

kolyshkin commented:
> LGTM, Are the SELinux tests running on Ubuntu now?

No :(

We need to switch to Fedora, I filed https://github.com/opencontainers/selinux/issues/220 for that.

This PR still makes sense but with or without it we actually don't test much :(

alexellis commented:
Current generations of arm64 servers available via cloud do not have working nested virtualisation. It may come in the future, where a standard runner could pivot into a RPM-based distro.

So whilst that's not available, and GitHub may not have it on their radar to support SELinux, we could provide SELinux based runners with fedora or a fedora-like distro like Alma for x86_64 and arm64.

Feel free to reach out to me to discuss options if you know of an end-user company willing to sponsor it.

kolyshkin commented:
@rhatdan @thaJeztah PTAL (the lack of selinux testing is now addressed by #221).

rhatdan commented:
It would be helpful if the audit.log is collected on failures, to see if there is an AVC reported.

kolyshkin commented:
> It would be helpful if the audit.log is collected on failures, to see if there is an AVC reported.

If we're talking about TestKeyLabel failures, I can easily reproduce it locally on Fedora 41 and I never see any new denials entries in audit.log. But let's discuss it here: https://github.com/opencontainers/selinux/issues/222.

If we're talking about this PR, is it sufficient to collect the output of `sudo ausearch -m AVC` after the tests? If yes, I just added this step as a last commit in this PR. If no, please suggest an alternative.

rhatdan commented on .github/workflows/validate.yml:
```diff
@@ -125,11 +125,16 @@ jobs:
         run: lima make -C /tmp/selinux GOARCH=386 test
 
+      - name: "Show AVC denials"
+        run: lima sudo ausearch -m AVC || true
```
-m AVC,USER_AVC

kolyshkin replied:
done

kolyshkin commented:
> If you run the test locally in permissive mode, does it still blow up? If yes, then the issue is DAC, if no, then the issue is SELinux, and the AVC must be blocked via a dontaudit rule.

Either `permissive` or `enforcing`, it blows up the same way:
<details>

```console
[kir@kir-tp1 selinux]$ sestatus
SELinux status:                 enabled
SELinuxfs mount:                /sys/fs/selinux
SELinux root directory:         /etc/selinux
Loaded policy name:             targeted
Current mode:                   enforcing
Mode from config file:          permissive
Policy MLS status:              enabled
Policy deny_unknown status:     allowed
Memory protection checking:     actual (secure)
Max kernel policy version:      33

[kir@kir-tp1 selinux]$ go test -count 100 -run KeyLabel ./go-selinux/label/ | grep -c FAIL
0

[kir@kir-tp1 selinux]$ go test -count 10000 -run KeyLabel ./go-selinux/label/ | grep -c FAIL
6719

[kir@kir-tp1 selinux]$ sudo setenforce 0
[sudo] password for kir: 

[kir@kir-tp1 selinux]$ go test -count 10000 -run KeyLabel ./go-selinux/label/ | grep -c FAIL
5912

[kir@kir-tp1 selinux]$ go test -count 10000 -run KeyLabel ./go-selinux/label/ | grep -c FAIL
6788

[kir@kir-tp1 selinux]$ sestatus 
SELinux status:                 enabled
SELinuxfs mount:                /sys/fs/selinux
SELinux root directory:         /etc/selinux
Loaded policy name:             targeted
Current mode:                   permissive
Mode from config file:          permissive
Policy MLS status:              enabled
Policy deny_unknown status:     allowed
Memory protection checking:     actual (secure)
Max kernel policy version:      33

```
</details>

rhatdan commented:
So this is a DAC issue, and the key is not writable from a non-root user, would be my guess.

kolyshkin commented:
> So this is a DAC issue, and the key is not writable from a non-root user, would be my guess.

@rhatdan this way it would *always* fail, but what we see is it fails randomly.

Also, it fails in the same way (i.e. randomly) when the test run as a root user (with `CAP_DAC_OVERRIDE` et al):
```console
[kir@kir-tp1 selinux]$ pwd
/home/kir/git/opencontainers/selinux
[kir@kir-tp1 selinux]$ cd go-selinux/label/
[kir@kir-tp1 label]$ go test -c .
[kir@kir-tp1 label]$ sudo ./label.test -test.run KeyLabel -test.count 10000 | grep -c FAIL
6205
```

(note I'm running 10000 iterations of the test, of which it fails 6205 times).

A (presumably) similar C program do not fail.

rhatdan commented:
Ok @kolyshkin has diagnosed this to be a GOLang thread issue.

alexellis commented:
I'm going to unsubscribe to this thread, but if anyone wants to take us up on the offer to add an SELinux/RPM-based distro as a runner, please reach out via email.

kolyshkin commented:
Flakes are being fixed by #227; I'd rather merge this PR first though (before #227).

thaJeztah commented on go-selinux/selinux_linux.go:
```diff
@@ -639,21 +639,6 @@ func (m mlsRange) String() string {
 }
 
-// TODO: remove min and max once Go < 1.21 is not supported.
-func max(a, b uint) uint {
```
Was there a strong reason not to continue supporting older Go versions? If not, I'd prefer continuing compatibility; we can either move this to a `//go:build ...` file, or rename the func to not shadow the builtin

kolyshkin replied:
No strong reasons; the weak ones are quite usual:
 - by switching to a newer version we can use new language and stdlib features, some of those I really like;
 - Go < 1.22 is not supported and might have some unpatched issues.

Sure, if there is a need, we can try maintaining backward compatibility, especially since this repo is not too bif or complex.

kolyshkin replied:
> we can either move this to a //go:build ... file

It's not possible. Now I remember I actually tried that before, only to get something like this:

```console
$ make
GOOS=linux GOARCH=amd64 go build  ./...
# github.com/opencontainers/selinux/go-selinux
go-selinux/selinux_linux.go:666:22: built-in max requires go1.21 or later (-lang was set to go1.19; check go.mod)
go-selinux/selinux_linux.go:669:23: built-in min requires go1.21 or later (-lang was set to go1.19; check go.mod)
```

kolyshkin replied:
Would like to hear your reasons for supporting Go 1.19.

thaJeztah commented on go.mod:
```diff
@@ -1,5 +1,5 @@
 module github.com/opencontainers/selinux
 
-go 1.19
+go 1.21
```
Same here, unless we really cannot

thaJeztah commented on go-selinux/selinux_linux.go:
```diff
@@ -138,7 +138,7 @@ func verifySELinuxfsMount(mnt string) bool {
 	}
 
-	if uint32(buf.Type) != uint32(unix.SELINUX_MAGIC) {
+	if uint32(buf.Type) != uint32(unix.SELINUX_MAGIC) { //nolint:gosec
```
Can you use the `#nosec G115` style? That way we can supported only `G115` without any other gosec issue that may be found

kolyshkin replied:
Alas this is not working for me. Perhaps you only used standalone `gosec`?

```console
$ golangci-lint run ./...
go-selinux/selinux_linux.go:141:11: G115: integer overflow conversion int64 -> uint32 (gosec)
	if uint32(buf.Type) != uint32(unix.SELINUX_MAGIC) { // #nosec G115 -- this code is right.
	         ^
```

kolyshkin replied:
This is weird. The same annotation works just fine in a different place:

```go
i0 := int(c.TrailingZeroBits()) //#nosec G115 -- don't expect TralingZeroBits to return values with highest bit set.
```

bit not here. I am lost 😿

kolyshkin replied:
OK, it doesn't work because it's after `{`. It works if I put in on a previous line:

```go
       //#nosec G115 -- there is no overflow here.
       if uint32(buf.Type) != uint32(unix.SELINUX_MAGIC) {
```

kolyshkin replied:
But, in this case it might want for the whole block, rather than just this line. I'm lost again 😕

kolyshkin replied:
Ah, the "whole block" is `return false` so I guess it's fine.

thaJeztah replied:
Yeah, they are kinda horrible, the `#nosec` comments! I think it's good to use them though (to not accidentally mask other security-related discoveries).

Thanks!

thaJeztah commented on go-selinux/selinux_linux.go:
```diff
@@ -583,7 +583,8 @@ func bitsetToStr(c *big.Int) string {
 
 	length := 0
-	for i := int(c.TrailingZeroBits()); i < c.BitLen(); i++ {
+	i0 := int(c.TrailingZeroBits()) //nolint:gosec
```
Same here, and possible add a comment after the `// #nosec G115 -- <short commend here >`

kolyshkin replied:
Done

kolyshkin commented:
@thaJeztah please take another look, I think I've addressed all your comments.

Aside from changes that you've requested, I added Go 1.19 to CI matrix, and added a new commit which removes a couple of unneeded `nolint:gosec` annotations.

thaJeztah commented:
Thanks! Yeah, users of older go versions should become rare, but as long as it's not adding a ton of burden (generally, this repository won't require too many updates I expect), I think  it's good to keep `<oldest supported>`, `current stable`, `old stable` as our test-matrix.

But we sure can update `<oldest supported>` once things become complicated / we _significantly_ benefit from new go features.

thaJeztah reviewed (approved):
LGTM, thanks!

kolyshkin commented:
> Thanks! Yeah, users of older go versions should become rare, but as long as it's not adding a ton of burden (generally, this repository won't require too many updates I expect), I think it's good to keep `<oldest supported>`, `current stable`, `old stable` as our test-matrix.

Yes to that!

> 
> But we sure can update `<oldest supported>` once things become complicated / we _significantly_ benefit from new go features.

In this particular case this was min/max functions, which I had to adopt in two regards:
 - change from uint to int (as their user changed the type);
 - change their names (or add linter annotations).
 
 I played a bit and decided the best way forward is to just bump min Go version and drop them.
 
 But it's not a big deal to keep compatibility if needed. At least for now.

rhatdan commented:
Might be worth adding renovate to this repo to keep the repo more up 2 date.

thaJeztah commented:
> Might be worth adding renovate to this repo to keep the repo more up 2 date.

For github actions used, and go version, probably. For go.mod, I think we can stick to go's MVS recommendations (specify the minimum version required)
